internal/jujuclient/dial.go

```diff
@@ -54,36 +57,116 @@ type Dialer struct {
 	JWTService                 *jimmjwx.JWTService
 }
 
-func (d *Dialer) createLoginRequest(ctx context.Context, ctl *dbmodel.Controller, modelTag names.ModelTag, p map[string]string) (*jujuparams.LoginRequest, error) {
-	// JIMM is automatically given all required permissions
-	permissions := p
-	if permissions == nil {
-		permissions = make(map[string]string)
-	}
-	permissions[ctl.ResourceTag().String()] = "superuser"
-	if modelTag.Id() != "" {
-		permissions[modelTag.String()] = "admin"
+func (d *Dialer) createLoginRequest(ctx context.Context, user *openfga.User, ctl *dbmodel.Controller, modelTag names.ModelTag, requirePermissionChecks map[string]string) (*jujuparams.LoginRequest, error) {
+
+	userPermissions := make(map[string]string)
+	if user == nil {
+		// if the user is not specify, we log in as the "admin" user
+		user = &openfga.User{
+			Identity: &dbmodel.Identity{
+				Name: "admin",
+			},
+		}
+
+		// if logging in as admin, we automatically grant the admin
+		// user all required permissions
+		if requirePermissionChecks != nil {
+			userPermissions = requirePermissionChecks
+		}
+
+		// we also state that the "admin" user is
+		// - controller superuser
+		// - model admin
+		userPermissions[ctl.ResourceTag().String()] = "superuser"
+		if modelTag.Id() != "" {
+			userPermissions[modelTag.String()] = "admin"
+		}
+	} else {
+		// if we are logging in as any other user, we need to check permissions
```

Comment:
Why are we checking permissions so late, at the dial stage? In the business logic we are already checking if the user has access then adding it to the list of permissions required when dialing, it seems wrong to check again.
The logic also seems to be too forgiving, allowing the caller to provide a resource tag and an access level but optionally allows the access level to be empty. I would argue we should simplify this (if we need to do this) and the access level should never be empty.